### ISSUE TITLE
[json] schema path not resolved properly for multi-root workspaces

### ISSUE THREAD

**Issue by schperplata** [microsoft/vscode#124491]
Issue Type: <b>Bug</b>

I have a multi-root VSC workspace with three folders:
* A/
* ./subfolderOfA
* `<external abs path>`

In A/, there is a `.code-workspace` file and in there, there are the following settings:
```
"files.associations": {
    "*.rq*": "json"
},
"json.schemas": [
    {
        "fileMatch": [
            "*.rqw"
        ],
        "url": "./samples/schemas/rqwSchema.json"
    }
]
```

When I edit some `.rqx` file somewhere inside A/, it seems like this JSON schema file path is only properly resolved If I remove folders ` ./subfolderOfA` and `<external abs path>` from my multi-root workspace. Shouldn't this work?

VS Code version: Code 1.56.2 (054a9295330880ed74ceaedda236253b4f39a335, 2021-05-12T17:13:13.157Z)
OS version: Windows_NT x64 10.0.19042

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz (8 x 2712)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.92GB (8.65GB free)|
|Process Argv|--crash-reporter-id 919e0ed9-46f3-4ad4-9e28-f35755bf8266|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (10)</summary>

Extension|Author (truncated)|Version
---|---|---
bracket-pair-colorizer-2|Coe|0.2.0
xml|Dot|2.5.1
tortoise-svn|fan|0.1.1
python|ms-|2020.9.114305
vscode-pylance|ms-|2021.5.3
jupyter|ms-|2021.6.832593372
cpptools|ms-|1.3.1
code-spell-checker|str|1.10.2
open-in-browser|tec|2.0.0
vscodeintellicode|Vis|1.2.14

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
pythonvspyt602:30300191
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492:30256859
pythonvspyt639:30300192
pythontb:30283811
pythonvspyt551:30291414
vspre833cf:30267465
pythonptprofiler:30281270
vshan820:30294714
vscorecescf:30290706
pythondataviewer:30285071
vscus158cf:30286554
vscgsv2:30301612
vscorehov:30302759
bridgeflight:30302069
vscod805:30301674

```

</details>

**Comment by marcnee** [comment]
Hi @aeschli 
Unfortunately, I am not familiar with your planning process but I wanted to ask if this bug will be solved in the near future?

**Comment by aeschli** [comment]
Sorry, I don't know when I will find time for looking at this.

**Comment by MeckPetarda** [comment]
It's honestly quite a shame that this feature is missing. Unless you want to modify the json file structure to include the "$schema" property you are stuck having to open the given sub-folders directly which is a significant detrement to the DX when working in multi-root workspaces.

**Comment by connor4312** [comment]
It seems like relative schema path resolution does not work on Windows, even when not in a multi-root workspace. Schemas always try to resolve in `C:/` rather than the workspace folder.

![image](https://user-images.githubusercontent.com/2230985/220718364-169c15d9-4a6d-4897-b54b-500d0cbbdb81.png)

**Comment by aeschli** [comment]
Note: relative path can not be used in user settings. But they should work in workspace files, where I could reproduce the problem, thanks @connor4312 !

**Comment by rzhao271** [comment]
Now I get `Unable to load schema from 'c:\Users\raymondzhao\work\emmet-work\testworkspace3': EISDIR: illegal operation on a directory, read.` in the Problems view with the following setup:

In a non-multi-root workspace:
`.vscode/settings.json`
```
    "json.schemas": [
        {
            "fileMatch": [
                "*.json"
            ],
            "url": "./catboost_info/la.json"
        }
    ]
```

`catboost_info/la.json`
```
{
}
```

`myfile.json`
```
{
	"$schema": ""
}
```

**Comment by aeschli** [comment]
I created a separate issue to investigate....

**Comment by aeschli** [comment]
To verify:

- clone https://github.com/aeschli/test-schema-in-workspace
- open workspace `test-schema-in-workspace/test-schema.code-workspace`
- look at the workspace settings: all files `ws.json` are associated with schema `./folder2/ws-schema.json`. Note that this is a relative path to the workspace file
- open both files `ws.json` and check that code completion gives back property `ws-folder2`. You can also click on the JSON language selector (`{ }` next to JSON in the status bar) to see the schema that is associated
- look at the folder setings:  `foo.json` is associated with schema `./folder1/foo-schema.json`. Note that this is a relative path to the folder
- open both files `foo.json` files and check that code completion gives back property `foo-folder1`, but only when in folder1. You can also click on the JSON language selector (`{ }` next to JSON in the status bar) to see the schema that is associated

**Comment by dbaeumer** [comment]
I tried to verify this using build

```
Version: 1.76.0-insider (user setup)
Commit: 92da9481c0904c6adfe372c12da3b7748d74bdcb
Date: 2023-02-28T11:50:30.886Z
Electron: 19.1.11
Chromium: 102.0.5005.196
Node.js: 16.14.2
V8: 10.2.154.26-electron.0
OS: Windows_NT x64 10.0.22623
Sandboxed: Yes
```

but the schemas are not used. Here is what I see in the UI:

<img width="706" alt="image" src="https://user-images.githubusercontent.com/1931590/221920989-fd4b2d57-adc7-42fe-98dd-3e9715eba9ce.png">

@aeschli please advise which build I should use for verification.

**Comment by aeschli** [comment]
Looks like it is still not working on Windows due to to an URI encoding issue.

**Comment by alx9r** [comment]
I tested as described in [the "test-schema-in-workspace" repro](https://github.com/microsoft/vscode/issues/124491#issuecomment-1448413811) with VS Code 1.76.0 on Windows.  

I observed the following for `foo.json` and `bar.json` in VS Code:

- no schema-defined code completion is offered
- there is no schema offered when clicking on the language selector

I then tested folder-only settings by opening VS Code to `folder1`.  I observed the same as above for `foo.json`.

FWIW, [this stackoverflow question](https://stackoverflow.com/questions/75664788/how-can-i-get-filematch-validation-working-while-editing-for-local-json-schemas) describes another repro with only folder settings (not workspace settings).  

Is there a known workaround to get validation working using `fileMatch` on Windows for 1.76.0?
